### ISSUE TITLE
Futures being awaited by a coroutine task should be cancelled if the task is cancelled

### ISSUE THREAD

**Issue by el-dee** [panda3d/panda3d#1136]
## Description

With asyncio, a when a coroutine task awaiting on a future is cancelled, the future is also cancelled (This is done internally by calling the `cancel()` method on the future from the asyncio task, see `asyncio.tasks.Task`)

A similar behaviour should be implemented in Panda3D for the awaitable requests like `tex.prepare()` call, an interval, or a `ModelLoadRequest`.  Note that a future might decide to ignore the cancel request and keep running, however its result will simply be ignored.

## Use Case

There are several use cases a coroutine task is awaiting a time consuming future or request (like loading a model or waiting for a concurrent thread to finish its work) and the purpose of the task is no longer relevant to the game (e.g. the object being loaded is no longer required or the current scene has been cleared) Not being able to cancel them might cause a pile up of request being performed uselessly.

See this thread for more background : https://discourse.panda3d.org/t/cant-cancel-an-awaiting-task/27458

**Comment by rdb** [comment]
Agreed.  This would allow cancelling eg. a whole loading sequence, cutscene or transition at once.  This would require a task to keep track of which future it is currently awaiting.

If we do this, we also need to introduce at the same time the equivalent of `asyncio.shield` to shield an awaitable from being affected by the cancellation of the surrounding task (eg. when awaiting a future that was not generated within the task).

Just thinking out loud, but I also wonder if similar to `cancel()` we could have a `finish()` that completes as normal but immediately skips any awaited intervals or animations to the end.  This might be useful if the user wants to skip through a cutscene or transition, such as to skip a scrolling text effect in a character dialogue window, in which case successive code should still run but complete in finite time.

**Comment by rdb** [comment]
Implemented.  Cancelling a task will cancel the future in it.  I added a `shield()` function to override this behaviour.

Since Panda's implementation is thread-safe (unlike asyncio), it also works with threading: if you're using `fut.result()` inside a threaded task to block the thread until the future is done, then cancelling the task will also cancel that future.

Testing would be welcome.  I added a bunch of unit tests but I am not fully confident that all corner cases have been explored.
